Core: Don't reset snapshotLog in `TableMetadata.removeRef` method

ebyhr opened the pull request:
Fixes #11777

nastra commented on core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java:
```diff
@@ -548,7 +552,8 @@ private static MetadataUpdate readSetSnapshotRef(JsonNode node) {
   private static MetadataUpdate readRemoveSnapshotRef(JsonNode node) {
     String refName = JsonUtil.getString(REF_NAME, node);
-    return new MetadataUpdate.RemoveSnapshotRef(refName);
+    boolean purge = JsonUtil.getBool(PURGE, node);
```
older clients won't send this flag, so we shouldn't be failing here

ebyhr replied:
Updated. The new logic sets `true` by default for backward compatibility.

ebyhr commented on core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java:
```diff
@@ -417,6 +420,7 @@ private static void writeSetSnapshotRef(MetadataUpdate.SetSnapshotRef update, Js
   private static void writeRemoveSnapshotRef(
       MetadataUpdate.RemoveSnapshotRef update, JsonGenerator gen) throws IOException {
     gen.writeStringField(REF_NAME, update.name());
+    gen.writeBooleanField(PURGE, update.purge());
```
Thanks for your suggestion. Posted https://lists.apache.org/thread/7q4rplst9tf9fpm7b9cco82j5n9q1hl7

amogh-jahagirdar reviewed (commented):
@ebyhr I think we may want to take a different approach. 

Currently, we clear the snapshot log when TableMetadata#removeRef is called [here](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1289) . If you see, the non-REST case goes through `resetMain` which won't clear the snapshot log but for the REST changes that are produced, we produce a RemoveSnapshotRef update type which when applied server side clears the log. 

Quite simply, I think we probably just want to remove the `snapshotLog.clear()` call in the removeRef API. In the past, I think there was concern that a user using that API may somehow rely on that behavior but really the only action on an Iceberg table that should remove history is snapshot expiration. `removeRef(MAIN)` should not really change history in anyway, this was probably just an oversight in the implementation done a long time ago. 

I also discussed this issue with a few others @Fokko @nastra @rdblue @danielcweeks and I think we're on the same page there. 

What are your thoughts? If you agree, would be happy to review the PR with the change to stop clearing the snapshot log, and some tests to verify that history is preserved during a REPLACE, and we can get a fix for this issue in!

ebyhr commented:
@amogh-jahagirdar There is no reason to disagree from my perspective. I was simply trying to keep the existing behavior because I didn't know the historical reason. Reverted most changes and removed `snapshotLog.clear()` call with some tests.

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:
```diff
@@ -2418,6 +2421,34 @@ public void testPaginationForListTables(int numberOfItems) {
   }
 
+  @Test
+  public void testReplaceTableKeepsSnapshotLog() {
```
Could we put this test in `CatalogTests`? That'll cover both REST + non-REST cases (TestRESTCatalog extends `CatalogTests`).

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:
```diff
@@ -2418,6 +2421,34 @@ public void testPaginationForListTables(int numberOfItems) {
+    table.refresh();
+    assertThat(((BaseTable) table).operations().current().snapshotLog())
+        .hasSize(2)
+        .containsAll(snapshotLogBeforeReplace);
```
Could we assert the exact snapshot log contents instead of just doing a containsAll check? We should have both the log before the replace and the new snapshot produced as part of the replace.

ebyhr commented:
@amogh-jahagirdar Thanks for your review. Addressed comments.

Fokko reviewed (approved):
Thanks for looking into this @ebyhr and @amogh-jahagirdar, I agree that this is the right approach.
